### ISSUE TITLE
Usage of balanceSequences and database transaction in certain scenario locks database queries

### ISSUE THREAD

**Issue by SargeKhan** [LiskHQ/lisk#1702]
### Expected behavior
Sequences should run exclusively and always resolve/reject.
### Actual behavior
The balanceSequences and database tx for applyBlock sometimes have a certain order which causes them to become codependent. This leads to balanceSequence callback waiting for the database tx to commit, and database tx is waiting for the (another) callback in balanceSequence to resolve.
### Scenario
Following are the conditions which reproduce the problem:
- Node has some transactions which update row for account A in the unconfirmed state. (Due to fillPool)
- Node starts trying to forge a block, so opens database tx
- `UndoUnconfirmedList` gets called as it's the first step of forge
- `UndoUnconfirmedList` calls `undoUnconfirm` on transaction T1
- We receive a new transaction T2 (which also effects account A) from public api, we add it to balanceSequence X
-  `undoUnconfirm` on transaction T1 resolves (which updated the row for account A), and we add it to balanceSequence Y.
- BalanceSequence start running (Order is X then Y), new transaction T2 received from public API calls `setAccountAndGet` for account A (which means it's updating a row which is already updated in undoUnconfirmList database tx).
- Now, this database query will wait for database tx to complete because database tx has rowExculsiveLock on that row.

And application will come to a standstill because inside Sequence X we are waiting for database Tx to finish and the database Tx will only finish after we resolve Sequence Y (which is dependent on sequence X).
### Which version(s) does this affect? (Environment, OS, etc...)
1.0.0

**Comment by nazarhussain** [comment]
The easiest way to understand the scenario is this simplest code snippet. 

```
      describe('db tx locks', () => {
            it('exclusive row level locking with update', function *() {
                const account = accountFixtures.Account();
                yield db.accounts.insert(account);

                return yield db.tx('applyBock', function*(t) {

                    // SELECT - Done
                    yield t.accounts.list({address: account.address});

                    // UPDATE - Done
                    yield t.accounts.update(account.address, {publicKey: '22222222'});

                    // UPDATE - Wait until transaction is committed
                    yield db.accounts.update(account.address, {publicKey: '11111111'});
                });
            });
        });
```

The last yield is executing  waiting for the transaction to be committed, but as well blocking the transaction block to be committed. So both are waiting for each other to finish.

**Comment by karmacoma** [comment]
Related to: https://github.com/LiskHQ/lisk/issues/538

**Comment by vitaly-t** [comment]
The line looks wrong:

```js
 yield db.accounts.update(account.address, {publicKey: '11111111'});
```

- it tries to execute a database query outside of the transaction connection. Shouldn't it be `yield t.accounts.update(...` instead? ;)

**Comment by SargeKhan** [comment]
@vitaly-t This test was a scenario which highlights the issue we were having.

**Comment by vitaly-t** [comment]
Ok. Such code should not exist in the app, i.e. a transaction must not contain queries inside executed on a different connection context, unless it is there by design.

If it is there by design, i.e. necessary, the containing transaction can permit such things by becoming writable.

**Comment by nazarhussain** [comment]
@vitaly-t Yes we are working on it to avoid such scenarios. Currently we identified one, but there can be different scenarios with such usage of transaction. We are trying to look into usage of sequences in the system that can cause this problem.

**Comment by SargeKhan** [comment]
After a healthy discussion with @nazarhussain @4miners @MaciejBaj, we concluded that this issue occurs when we have multiple balanceSequences inside a database transaction.
The solution we came up with is that we should have at most one balanceSequence in a database transaction. This means that, in applyBlock function, applyUnconfirmed for all transactions should be wrapped inside one balanceSequence. (After #1652 is fixed)

Another good suggestion from the meeting was that we should wrap `undoUnconfirmedList`, `applyUnconfirmedList` inside one balanceSequence, as opposed to our current implementation, where we add a new balanceSequence for each transaction.
